packages/core/parcel-bundler/src/scope-hoisting/renamer.js

```diff
@@ -3,31 +3,7 @@ function rename(scope, oldName, newName) {
     return;
   }
 
-  let binding = scope.getBinding(oldName);
-
-  // Rename all constant violations
-  for (let violation of binding.constantViolations) {
-    let bindingIds = violation.getBindingIdentifierPaths(true, false);
-    for (let name in bindingIds) {
-      if (name === oldName) {
-        for (let idPath of bindingIds[name]) {
-          idPath.node.name = newName;
-        }
-      }
-    }
-  }
-
-  // Rename all references
-  for (let path of binding.referencePaths) {
-    if (path.node.name === oldName) {
-      path.node.name = newName;
-    }
-  }
-
-  // Rename binding identifier, and update scope.
-  binding.identifier.name = newName;
-  scope.bindings[newName] = binding;
-  delete scope.bindings[oldName];
+  scope.rename(oldName, newName);
```

Comment:
Unfortunately the builtin renamer in babel is very slow, which is why we made our own. The builtin one does a tree traverse, so when bulk renaming variables it is O(n^2).

Could you see if it's possible to fix the bug by modifying the renamer we have already in parcel rather than switching back to the slow babel one?

Reply:
Somehow this isn't consistent. First, the AST contains the class transpiled to a function. Then, inside the renaming for-loop it's a real classdeclaration again (correctly renamed) and afterwards, it a transpiled class again (but the `console.log(Component)` worked).
`rename` is all called by a ImportDeclaration visitor, could the class transform somehow run in parallel?

Replacing the `for (let path of binding.referencePaths)`-part with babel's solution works fine:
```js
scope.traverse(scope.block, {
	ReferencedIdentifier({node}) {
		if (node.name === oldName) {
			node.name = newName;
		}
	}
});
```


```js
function rename(scope, oldName, newName) {
	// -------- AST of scope -------- (correct)
	import Component from "./x.js";
	console.log(Component);

	var App =
	/*#__PURE__*/
	function (_Component) { // <--------------------------------------- class transpiled
	  _inherits(App, _Component);

	  function App() {
	    _classCallCheck(this, App);

	    return _possibleConstructorReturn(this, _getPrototypeOf(App).apply(this, arguments));
	  }

	  return App;
	}(Component);
	// -------- AST --------

	const binding = scope.getBinding(oldName);

	for (let path of binding.referencePaths) {
		if (path.node.name === oldName) {
			update name

			// -------- AST of path.parent -------- (1st run)
			console.log($Focm$import$Component)
			// -------- AST --------

			// -------- AST of path.parent -------- (2nd run)
			class App extends $Focm$import$Component { // <------------- now a classdeclaration again?
			  constructor() {
			    return _possibleConstructorReturn(this,
			    			_getPrototypeOf(App).apply(this, arguments));
			  }
			}
			// -------- AST --------
		}


	// -------- AST of scope -------- (halfway correct)
	import Component from "./x.js";
	console.log($Focm$import$Component); // <---------------------------- worked

	var App =
	/*#__PURE__*/
	function (_Component) {
	  _inherits(App, _Component);

	  function App() {
	    _classCallCheck(this, App);

	    return _possibleConstructorReturn(this, _getPrototypeOf(App).apply(this, arguments));
	  }

	  return App;
	}(Component); // <--------------------------------------------------- ?????
	// -------- AST --------
}
```

Reply:
Well, this would fix it as well - but I guess crawl isn't much better than traverse?

```diff
diff --git a/packages/core/parcel-bundler/src/scope-hoisting/hoist.js b/packages/core/parcel-bundler/src/scope-hoisting/hoist.js
index afc01e4..70b84a5 100644
--- a/packages/core/parcel-bundler/src/scope-hoisting/hoist.js
+++ b/packages/core/parcel-bundler/src/scope-hoisting/hoist.js
@@ -344,6 +344,8 @@ module.exports = {
   ImportDeclaration(path, asset) {
     // For each specifier, rename the local variables to point to the imported name.
     // This will be replaced by the final variable name of the resolved asset in the packager.
+    path.scope.crawl();
+
     for (let specifier of path.node.specifiers) {
       let id = getIdentifier(asset, 'import', specifier.local.name);
       rename(path.scope, specifier.local.name, id.name);
```

Reply:
could you try putting that `scope.crawl()` call at the top in the `Program.enter` visitor? This way it will happen once at the beginning of the traversal rather than every time an import declaration is hit.

Reply:
Still works